fluent-react/src/localization.ts

```diff
@@ -73,6 +79,195 @@ export class ReactLocalization {
+
+    let localizedProps: Record<string, string> | undefined;
+    // The default is to forbid all message attributes. If the attrs prop exists
+    // on the Localized instance, only set message attributes which have been
+    // explicitly allowed by the developer.
+    if (args?.attrs && msg.attributes) {
+      localizedProps = {};
+      errors = [];
+      for (const [name, allowed] of Object.entries(args?.attrs)) {
+        if (allowed && name in msg.attributes) {
+          localizedProps[name] = bundle.formatPattern(
+            msg.attributes[name],
+            args?.vars,
+            errors
+          );
+        }
+      }
+      for (let error of errors) {
+        this.reportError(error);
+      }
+    }
+
+    // If the component to render is a known void element, explicitly dismiss the
+    // message value and do not pass it to cloneElement in order to avoid the
+    // "void element tags must neither have `children` nor use
+    // `dangerouslySetInnerHTML`" error.
+    if (typeof componentToRender.type === "string" && componentToRender.type in voidElementTags) {
+      return cloneElement(componentToRender, localizedProps);
+    }
+
+    // If the message has a null value, we're only interested in its attributes.
+    // Do not pass the null value to cloneElement as it would nuke all children
+    // of the wrapped component.
+    if (msg.value === null) {
+      return cloneElement(componentToRender, localizedProps);
+    }
+
+    errors = [];
+    const messageValue = bundle.formatPattern(msg.value, args?.vars, errors);
+    for (let error of errors) {
+      this.reportError(error);
+    }
+
+    // If the message value doesn't contain any markup nor any HTML entities,
+    // insert it as the only child of the component to render.
+    if (!reMarkup.test(messageValue) || this.parseMarkup === null) {
+      return cloneElement(componentToRender, localizedProps, messageValue);
+    }
+
+    let elemsLower: Record<string, ReactElement>;
+    if (args?.elems) {
+      elemsLower = {};
+      for (let [name, elem] of Object.entries(args?.elems)) {
+        elemsLower[name.toLowerCase()] = elem;
+      }
+    }
+
+    // If the message contains markup, parse it and try to match the children
+    // found in the translation with the args passed to this function.
+    const translationNodes = this.parseMarkup(messageValue);
+    const translatedChildren = translationNodes.map(childNode => {
+      if (childNode.nodeName === "#text") {
+        return childNode.textContent;
+      }
+
+      const childName = childNode.nodeName.toLowerCase();
+
+      // If the child is not expected just take its textContent.
+      if (
+        !elemsLower ||
+        !Object.prototype.hasOwnProperty.call(elemsLower, childName)
+      ) {
+        return childNode.textContent;
+      }
+
+      const sourceChild = elemsLower[childName];
+
+      // Ignore elems which are not valid React elements.
+      if (!isValidElement(sourceChild)) {
```

Comment:
Nitpick: It's a bit odd to do this check only now rather than when building `elemsLower`.

Reply:
Hmm, I'm not entirely sure how this code works (it also [already existed](https://github.com/projectfluent/fluent.js/blob/8c2eae431ef89f22d6d07f860e5043fe4eac3fc3/fluent-react/src/localized.ts#L200)), but I think "ignore" here means "preserve, but do not translate" rather than "filter out"?

Reply:
I meant that this check could/should be somewhere around line 220 instead, so that only valid elements are ever added into `elemsLower`.

Reply:
Ah gotcha. b950f1c7